{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "14e6ae0a-3657-431c-8b5a-21d1fb042bc7",
   "metadata": {},
   "source": [
    "# \"Molecular Microbial Ecology\" MBGW2.2.3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "397ac7b0-afe8-40be-a1e4-0499018a622e",
   "metadata": {},
   "source": [
    "## Assessing bin completeness and contamination and winning insights into the taxonomy of your bins\n",
    "\n",
    "Now that we are done with binning our genomes, we want to know how “good” our genome bins are, with good meaning how complete and contaminated. `METAWRAP` already provided us with some information in that regard, we now want to take a closer look.\n",
    "\n",
    "Contamination can for instance arise if we extracted contigs that originated from different organisms. This is especially an issue when you are dealing with somewhat closely related organisms. How can we assess genome contamination and completeness? The answer are distinct sets of single-copy marker genes characteristic of taxonomic groups of interest. \n",
    "\n",
    "Dependent on how many of them we find in a putative genome bin we can draw conclusions about its completeness. At the same time, we can use them to deduce information about whether the genome is contaminated and to what extent. This is possible by considering the copy number of the single-copy marker genes. Basically, they should occur once, if this is not the case, if genes are present multiple times, the genome is presumably contaminated with foreign sequence information."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d4a93ebb-69b9-429f-8b3b-61b6db23c406",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>QUESTION/TASK:</b> \n",
    "What are single-copy core genes? can you think about obvious weaknesses of this approach?\n",
    "</div>\n",
    "\n",
    "For assessing the completeness/contamination and taxonomic affiliation we use [`checkM`]() and [`GTDB-Tk`](https://github.com/ecogenomics/gtdbtk).\n",
    "\n",
    "`GTDB-Tk` relies on the Genome Taxonomy Database [`GTDB`](https://gtdb.ecogenomic.org/) that is maintained by the Australian Center for Ecogenomics.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dd07c58e-283f-42d3-8620-e2448dfbd034",
   "metadata": {},
   "outputs": [],
   "source": [
    "### We move to our practical course directory and first activate the needed checkM conda environment\n",
    "cd ~/data/MBGW223\n",
    "conda activate checkm\n",
    "checkm lineage_wf ./02_ASSEMBLY_BINNING/bin_refinement_megahit_metawrap/metawrap_70_10_bins ./02_ASSEMBLY_BINNING/checkm_metawrap_refined -t 6 --reduced_tree\n",
    "checkm qa ./02_ASSEMBLY_BINNING/checkm_metawrap_refined/lineage.ms ./02_ASSEMBLY_BINNING/checkm_metawrap_refined -o 2 -f checkm_summary.tsv\n",
    "### Next, we want to classify (taxonomically place our bins) with GTDB-Tk\n",
    "conda deactivate\n",
    "conda activate gtdbtk_1.5\n",
    "gtdbtk classify_wf --genome_dir ./02_ASSEMBLY_BINNING/bin_refinement_megahit_metawrap/metawrap_70_10_bins --out_dir ./02_ASSEMBLY_BINNING/gtdbtk_metawrap_refined "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0546d46a-da2e-44e8-85ee-3f2557c916c1",
   "metadata": {},
   "source": [
    "Used parameters:\n",
    "\n",
    "`lineage_wf`: checkM comes with different workflows, lineage_wf is a workflow that assesses contamination and completeness based on lineage-specific sets of single copy core genes. checkM first identifies open reading frames, checks whether they code for single copy core genes, if yes these are extracted and matched against databases to infer the taxonomic identity of the respective genome. Based on the identified taxonomy, a set of single copy core genes is used to assess completeness/contamination.\n",
    "\n",
    "`--reduced_tree`: checkM eats memory like cookies, in order to reduce the computational burden we use a reduced tree in the context of assigning taxonomy.\n",
    "\n",
    "`qa`: This is basically the last step of the previous workflow that we repeat to generate text output that we can save. The workflow just shows the output in the commandline.\n",
    "\n",
    "`--genome_dir`: Path of the folder containing your (refined) MAGs.\n",
    "\n",
    "`--out_dir`: Path to the location where gtdbtk will save its output.\n",
    "\n",
    "`GTDB-Tk` is classifying genomes combining three approaches:\n",
    "\n",
    "1. Placing them in a reference tree based on a defined set of single-copy-marker genes.\n",
    "2. Calculating ANIs for a collection of reference genomes.\n",
    "3. Determining relative evolutionary divergence.\n",
    "\n",
    "`GTDB-Tk` uses big reference databases, meaning the run takes a while. Once it is done, check the output folder and download the summary files (also from `checkM`).\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>QUESTION/TASK:</b> \n",
    "<ul>\n",
    "  <li>How complete are our genomes?</li>\n",
    "  <li>What is the checkM output file telling us?</li>\n",
    "  <li>What taxa are we dealing with?</li>\n",
    "</ul>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "764a51e5-d726-4052-b098-0f912c6bbc15",
   "metadata": {},
   "source": [
    "## Genome annnotation and metabolic potential\n",
    "Genome annotation refers to the process of identifying coding (genes, ORFs), non-coding (rRNA, tRNA, siRNA, antisense RNA), as well as additional structural features (signal peptides, transmembrane helices, CRISPR arrays, etc.). Although available tools cover all of these aspects, manual curation of genome annotations is essential to yield a high quality genome annotation. \n",
    "\n",
    "### Genome annotation\n",
    "Obviously we do not have the time for extensive manual curation, instead we will use [`prokka`](https://github.com/tseemann/prokka) as tool for rapid prokaryotic genome annotation. `prokka` uses a lot of different tools and database resources and allows to annotate a typical genomes in a few minutes. prokka has a number of optional parameters, but we will call it in a very basic way.\n",
    "\n",
    "<img src=\"img/prokka.png\" alt=\"the prokka pipeline\" width=\"500\"/>\n",
    "\n",
    "<font size=\"2\"> Prokka pipeline, incl. tools and dbs. </font>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e9dc4c5-3947-4eb4-9723-2dd9cbf081db",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[18:32:34] This is prokka 1.14.6\n",
      "[18:32:34] Written by Torsten Seemann <torsten.seemann@gmail.com>\n",
      "[18:32:34] Homepage is https://github.com/tseemann/prokka\n",
      "[18:32:34] Local time is Thu Aug 24 18:32:34 2023\n",
      "[18:32:34] You are lu87neb\n",
      "[18:32:34] Operating system is linux\n",
      "[18:32:34] You have BioPerl 1.7.8\n",
      "Argument \"1.7.8\" isn't numeric in numeric lt (<) at /data/mambaforge/envs/genome_annot/bin/prokka line 259.\n",
      "[18:32:34] System has 48 cores.\n",
      "[18:32:34] Will use maximum of 8 cores.\n",
      "[18:32:34] Annotating as >>> Bacteria <<<\n",
      "[18:32:34] Creating new output folder: ./03_GENOM_ANNOT/bin.10_prokka\n",
      "[18:32:34] Running: mkdir -p \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\n",
      "[18:32:34] Using filename prefix: PROKKA_08242023.XXX\n",
      "[18:32:34] Setting HMMER_NCPU=1\n",
      "[18:32:34] Writing log to: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.log\n",
      "[18:32:34] Command: /data/mambaforge/envs/genome_annot/bin/prokka --outdir ./03_GENOM_ANNOT/bin.10_prokka --locustag bin.10 ./02_ASSEMBLY_BINNING/bin_refinement_spades_metawrap/metawrap_70_10_bins/bin.10.fa\n",
      "[18:32:34] Appending to PATH: /data/mambaforge/envs/genome_annot/bin\n",
      "[18:32:34] Looking for 'aragorn' - found /data/mambaforge/envs/genome_annot/bin/aragorn\n",
      "[18:32:34] Determined aragorn version is 001002 from 'ARAGORN v1.2.41 Dean Laslett'\n",
      "[18:32:34] Looking for 'barrnap' - found /data/mambaforge/envs/genome_annot/bin/barrnap\n",
      "[18:32:34] Determined barrnap version is 000009 from 'barrnap 0.9'\n",
      "[18:32:34] Looking for 'blastp' - found /data/mambaforge/envs/genome_annot/bin/blastp\n",
      "[18:32:34] Determined blastp version is 002013 from 'blastp: 2.13.0+'\n",
      "[18:32:34] Looking for 'cmpress' - found /data/mambaforge/envs/genome_annot/bin/cmpress\n",
      "[18:32:34] Determined cmpress version is 001001 from '# INFERNAL 1.1.4 (Dec 2020)'\n",
      "[18:32:34] Looking for 'cmscan' - found /data/mambaforge/envs/genome_annot/bin/cmscan\n",
      "[18:32:34] Determined cmscan version is 001001 from '# INFERNAL 1.1.4 (Dec 2020)'\n",
      "[18:32:34] Looking for 'egrep' - found /usr/bin/egrep\n",
      "[18:32:34] Looking for 'find' - found /usr/bin/find\n",
      "[18:32:34] Looking for 'grep' - found /usr/bin/grep\n",
      "[18:32:34] Looking for 'hmmpress' - found /data/mambaforge/envs/genome_annot/bin/hmmpress\n",
      "[18:32:34] Determined hmmpress version is 003003 from '# HMMER 3.3.2 (Nov 2020); http://hmmer.org/'\n",
      "[18:32:34] Looking for 'hmmscan' - found /data/mambaforge/envs/genome_annot/bin/hmmscan\n",
      "[18:32:34] Determined hmmscan version is 003003 from '# HMMER 3.3.2 (Nov 2020); http://hmmer.org/'\n",
      "[18:32:34] Looking for 'java' - found /data/mambaforge/envs/genome_annot/bin/java\n",
      "[18:32:34] Looking for 'makeblastdb' - found /data/mambaforge/envs/genome_annot/bin/makeblastdb\n",
      "[18:32:34] Determined makeblastdb version is 002013 from 'makeblastdb: 2.13.0+'\n",
      "[18:32:34] Looking for 'minced' - found /data/mambaforge/envs/genome_annot/bin/minced\n",
      "[18:32:35] Determined minced version is 004002 from 'minced 0.4.2'\n",
      "[18:32:35] Looking for 'parallel' - found /data/mambaforge/envs/genome_annot/bin/parallel\n",
      "[18:32:35] Determined parallel version is 20220722 from 'GNU parallel 20220722'\n",
      "[18:32:35] Looking for 'prodigal' - found /data/mambaforge/envs/genome_annot/bin/prodigal\n",
      "[18:32:35] Determined prodigal version is 002006 from 'Prodigal V2.6.3: February, 2016'\n",
      "[18:32:35] Looking for 'prokka-genbank_to_fasta_db' - found /data/mambaforge/envs/genome_annot/bin/prokka-genbank_to_fasta_db\n",
      "[18:32:35] Looking for 'sed' - found /usr/bin/sed\n",
      "[18:32:35] Looking for 'tbl2asn' - found /data/mambaforge/envs/genome_annot/bin/tbl2asn\n",
      "[18:32:35] Determined tbl2asn version is 025007 from 'tbl2asn 25.7   arguments:'\n",
      "[18:32:35] Using genetic code table 11.\n",
      "[18:32:35] Loading and checking input file: ./02_ASSEMBLY_BINNING/bin_refinement_spades_metawrap/metawrap_70_10_bins/bin.10.fa\n",
      "[18:32:35] Wrote 3 contigs totalling 1074755 bp.\n",
      "[18:32:35] Predicting tRNAs and tmRNAs\n",
      "[18:32:35] Running: aragorn -l -gc11  -w \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.fna\n",
      "[18:32:35] 1 tRNA-Ala c[22606,22679] 34 (tgc)\n",
      "[18:32:35] 2 tRNA-Ser c[106838,106927] 35 (tga)\n",
      "[18:32:35] 3 tRNA-Arg [107734,107808] 35 (gcg)\n",
      "[18:32:35] 4 tRNA-Ser [107813,107898] 35 (cga)\n",
      "[18:32:35] 5 tRNA-Arg [108001,108074] 34 (ccg)\n",
      "[18:32:35] 6 tRNA-Ser [108594,108678] 35 (gga)\n",
      "[18:32:35] 7 tRNA-Ser [110651,110743] 37 (gct)\n",
      "[18:32:35] 8 tRNA-Ile [146749,146821] 33 (tat)\n",
      "[18:32:35] 9 tRNA-Met c[148735,148803] 31 (cat)\n",
      "[18:32:35] 10 tRNA-Met [149022,149096] 35 (cat)\n",
      "[18:32:35] 11 tRNA-Asn [158372,158446] 33 (gtt)\n",
      "[18:32:35] 12 tRNA-Pro c[231923,231994] 33 (ggg)\n",
      "[18:32:35] 13 tRNA-Gln [387158,387229] 33 (ctg)\n",
      "[18:32:35] 14 tRNA-Glu c[423370,423444] 35 (ctc)\n",
      "[18:32:35] 15 tRNA-Val [551620,551694] 35 (cac)\n",
      "[18:32:35] 16 tRNA-Thr c[584076,584149] 34 (cgt)\n",
      "[18:32:35] 17 tRNA-Asp [607495,607569] 35 (gtc)\n",
      "[18:32:35] 18 tRNA-Gly [608840,608913] 33 (ccc)\n",
      "[18:32:35] 19 tRNA-Glu c[631808,631882] 35 (ttc)\n",
      "[18:32:35] 20 tRNA-Lys [632876,632949] 33 (ttt)\n",
      "[18:32:35] 21 tRNA-Lys c[648354,648427] 33 (ctt)\n",
      "[18:32:35] 22 tRNA-Arg c[648837,648911] 35 (tcg)\n",
      "[18:32:35] 23 tRNA-His c[662240,662314] 33 (gtg)\n",
      "[18:32:35] 24 tRNA-Arg [663608,663681] 34 (tct)\n",
      "[18:32:35] 25 tRNA-Gly c[666343,666414] 33 (tcc)\n",
      "[18:32:35] 26 tRNA-Pro c[669701,669774] 34 (tgg)\n",
      "[18:32:35] 27 tRNA-Leu c[669801,669886] 34 (taa)\n",
      "[18:32:35] 1 tRNA-Ala [13703,13774] 33 (cgc)\n",
      "[18:32:35] 2 tRNA-Ala [13844,13916] 33 (ggc)\n",
      "[18:32:35] 3 tRNA-Pro c[60968,61042] 33 (cgg)\n",
      "[18:32:35] 4 tRNA-Arg [61171,61243] 34 (cct)\n",
      "[18:32:35] 5 tRNA-Thr [104146,104219] 33 (tgt)\n",
      "[18:32:35] 6 tRNA-Tyr [104223,104309] 34 (gta)\n",
      "[18:32:35] 7 tRNA-Thr [104330,104402] 33 (ggt)\n",
      "[18:32:35] 8 tRNA-Trp [104904,104976] 34 (cca)\n",
      "[18:32:35] 9 tRNA-Ile c[130582,130655] 34 (gat)\n",
      "[18:32:35] 10 tRNA-Cys c[130707,130780] 34 (gca)\n",
      "[18:32:35] 11 tRNA-Ser c[196073,196158] 35 (act)\n",
      "[18:32:35] 12 tRNA-Leu c[289191,289276] 34 (caa)\n",
      "[18:32:35] 1 tRNA-Gln c[27954,28027] 33 (ttg)\n",
      "[18:32:35] 2 tRNA-Gly [31470,31542] 33 (gcc)\n",
      "[18:32:35] 3 tRNA-Phe [31583,31659] 36 (gaa)\n",
      "[18:32:35] 4 tRNA-Leu [32053,32136] 34 (cag)\n",
      "[18:32:35] 5 tRNA-Leu [32192,32277] 34 (gag)\n",
      "[18:32:35] 6 tRNA-Val [32278,32352] 35 (tac)\n",
      "[18:32:35] 7 tRNA-Leu [33968,34051] 34 (tag)\n",
      "[18:32:35] 8 tRNA-Val [38448,38526] 36 (gac)\n",
      "[18:32:35] Found 47 tRNAs\n",
      "[18:32:35] Predicting Ribosomal RNAs\n",
      "[18:32:35] Running Barrnap with 8 threads\n",
      "[18:32:36] 1 NODE_23_length_318163_cov_15.857520 127616 16S ribosomal RNA\n",
      "[18:32:36] 2 NODE_2_length_684733_cov_15.923411 18695 5S ribosomal RNA\n",
      "[18:32:36] 3 NODE_2_length_684733_cov_15.923411 19786 23S ribosomal RNA (partial)\n",
      "[18:32:36] Found 3 rRNAs\n",
      "[18:32:36] Skipping ncRNA search, enable with --rfam if desired.\n",
      "[18:32:36] Total of 50 tRNA + rRNA features\n",
      "[18:32:36] Searching for CRISPR repeats\n",
      "[18:32:36] Found 0 CRISPRs\n",
      "[18:32:36] Predicting coding sequences\n",
      "[18:32:36] Contigs total 1074755 bp, so using single mode\n",
      "[18:32:36] Running: prodigal -i \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.fna -c -m -g 11 -p single -f sco -q\n",
      "[18:32:39] Excluding CDS which overlaps existing RNA (tRNA) at NODE_2_length_684733_cov_15.923411:108015..108530 on + strand\n",
      "[18:32:39] Excluding CDS which overlaps existing RNA (tRNA) at NODE_2_length_684733_cov_15.923411:422969..423373 on + strand\n",
      "[18:32:39] Excluding CDS which overlaps existing RNA (tRNA) at NODE_2_length_684733_cov_15.923411:608862..609629 on + strand\n",
      "[18:32:39] Excluding CDS which overlaps existing RNA (tRNA) at NODE_2_length_684733_cov_15.923411:647813..648358 on + strand\n",
      "[18:32:39] Excluding CDS which overlaps existing RNA (tRNA) at NODE_173_length_71859_cov_15.907303:32067..32180 on + strand\n",
      "[18:32:39] Excluding CDS which overlaps existing RNA (tRNA) at NODE_173_length_71859_cov_15.907303:33961..34119 on + strand\n",
      "[18:32:39] Found 1083 CDS\n",
      "[18:32:39] Connecting features back to sequences\n",
      "[18:32:39] Not using genus-specific database. Try --usegenus to enable it.\n",
      "[18:32:39] Annotating CDS, please be patient.\n",
      "[18:32:39] Will use 8 CPUs for similarity searching.\n",
      "[18:32:40] There are still 1083 unannotated CDS left (started with 1083)\n",
      "[18:32:40] Will use blast to search against /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/IS with 8 CPUs\n",
      "[18:32:40] Running: cat \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.IS\\.tmp\\.1134873\\.faa | parallel --gnu --plain -j 8 --block 20440 --recstart '>' --pipe blastp -query - -db /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/IS -evalue 1e-30 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.IS\\.tmp\\.1134873\\.blast 2> /dev/null\n",
      "[18:32:42] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.IS.tmp.1134873.faa\n",
      "[18:32:42] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.IS.tmp.1134873.blast\n",
      "[18:32:43] There are still 1082 unannotated CDS left (started with 1083)\n",
      "[18:32:43] Will use blast to search against /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/AMR with 8 CPUs\n",
      "[18:32:43] Running: cat \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.AMR\\.tmp\\.1134873\\.faa | parallel --gnu --plain -j 8 --block 20429 --recstart '>' --pipe blastp -query - -db /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/AMR -evalue 1e-300 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.AMR\\.tmp\\.1134873\\.blast 2> /dev/null\n",
      "[18:32:45] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.AMR.tmp.1134873.faa\n",
      "[18:32:45] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.AMR.tmp.1134873.blast\n",
      "[18:32:45] There are still 1082 unannotated CDS left (started with 1083)\n",
      "[18:32:45] Will use blast to search against /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/sprot with 8 CPUs\n",
      "[18:32:45] Running: cat \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.sprot\\.tmp\\.1134873\\.faa | parallel --gnu --plain -j 8 --block 20429 --recstart '>' --pipe blastp -query - -db /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/sprot -evalue 1e-09 -qcov_hsp_perc 80 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.sprot\\.tmp\\.1134873\\.blast 2> /dev/null\n",
      "[18:32:56] Modify product: Probable peptidoglycan glycosyltransferase FtsW => putative peptidoglycan glycosyltransferase FtsW\n",
      "[18:32:56] Modify product: UPF0213 protein EF_2693 => hypothetical protein\n",
      "[18:32:56] Modify product: Probable murein peptide carboxypeptidase => putative murein peptide carboxypeptidase\n",
      "[18:32:56] Modify product: GTP-binding protein TypA/BipA homolog => GTP-binding protein TypA/BipA \n",
      "[18:32:56] Modify product: Uncharacterized protein Rv1405c => putative protein\n",
      "[18:32:56] Modify product: Probable N-acetyl-alpha-D-glucosaminyl L-malate deacetylase 2 => putative N-acetyl-alpha-D-glucosaminyl L-malate deacetylase 2\n",
      "[18:32:56] Modify product: Probable undecaprenyl-phosphate N-acetylglucosaminyl 1-phosphate transferase => putative undecaprenyl-phosphate N-acetylglucosaminyl 1-phosphate transferase\n",
      "[18:32:56] Modify product: Uncharacterized ABC transporter ATP-binding protein YknY => putative ABC transporter ATP-binding protein YknY\n",
      "[18:32:56] Modify product: Probable endonuclease 4 => putative endonuclease 4\n",
      "[18:32:56] Modify product: Uncharacterized protein YqeY => putative protein YqeY\n",
      "[18:32:56] Modify product: Ribonuclease TTHA0252 => Ribonuclease\n",
      "[18:32:56] Modify product: Probable copper-transporting ATPase PacS => putative copper-transporting ATPase PacS\n",
      "[18:32:56] Modify product: Uncharacterized glycosyltransferase sll0501 => putative glycosyltransferase\n",
      "[18:32:56] Modify product: Protease HtpX homolog => Protease HtpX \n",
      "[18:32:56] Modify product: Probable tRNA-dihydrouridine synthase => putative tRNA-dihydrouridine synthase\n",
      "[18:32:56] Modify product: Uncharacterized zinc protease Rv2782c => putative zinc protease\n",
      "[18:32:56] Modify product: Probable manganese efflux pump MntP => putative manganese efflux pump MntP\n",
      "[18:32:56] Modify product: UPF0047 protein Rv2556c => hypothetical protein\n",
      "[18:32:56] Modify product: Putative phosphoribosyl transferase Rv0571c => Putative phosphoribosyl transferase\n",
      "[18:32:56] Modify product: Probable polyglutamine synthesis accessory protein Rv0574c => putative polyglutamine synthesis accessory protein\n",
      "[18:32:56] Modify product: UPF0098 protein CT_736 => hypothetical protein\n",
      "[18:32:56] Modify product: Putative phosphoribosyl transferase Rv0571c => Putative phosphoribosyl transferase\n",
      "[18:32:56] Modify product: UPF0176 protein SP_0095 => hypothetical protein\n",
      "[18:32:56] Modify product: Uncharacterized RNA pseudouridine synthase Rv1540 => putative RNA pseudouridine synthase\n",
      "[18:32:56] Modify product: Probable transcriptional regulatory protein aq_1575 => putative transcriptional regulatory protein\n",
      "[18:32:56] Modify product: Probable multidrug ABC transporter permease YbhS => putative multidrug ABC transporter permease YbhS\n",
      "[18:32:56] Modify product: Bifunctional protein FolD => Bifunctional protein FolD protein\n",
      "[18:32:56] Modify product: Putative transport protein Rv0205 => Putative transport protein\n",
      "[18:32:56] Cleaned 28 /product names\n",
      "[18:32:56] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.sprot.tmp.1134873.faa\n",
      "[18:32:56] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.sprot.tmp.1134873.blast\n",
      "[18:32:56] There are still 719 unannotated CDS left (started with 1083)\n",
      "[18:32:56] Will use hmmer3 to search against /data/mambaforge/envs/genome_annot/db/hmm/HAMAP.hmm with 8 CPUs\n",
      "[18:32:56] Running: cat \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.HAMAP\\.hmm\\.tmp\\.1134873\\.faa | parallel --gnu --plain -j 8 --block 12691 --recstart '>' --pipe hmmscan --noali --notextw --acc -E 1e-09 --cpu 1 /data/mambaforge/envs/genome_annot/db/hmm/HAMAP.hmm /dev/stdin > \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.HAMAP\\.hmm\\.tmp\\.1134873\\.hmmer3 2> /dev/null\n",
      "[18:33:02] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.HAMAP.hmm.tmp.1134873.faa\n",
      "[18:33:02] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.HAMAP.hmm.tmp.1134873.hmmer3\n",
      "[18:33:02] Labelling remaining 631 proteins as 'hypothetical protein'\n",
      "[18:33:02] Possible /pseudo 'D-inositol-3-phosphate glycosyltransferase' at NODE_2_length_684733_cov_15.923411 position 492107\n",
      "[18:33:02] Possible /pseudo 'Nucleoside diphosphate kinase' at NODE_23_length_318163_cov_15.857520 position 56692\n",
      "[18:33:02] Found 369 unique /gene codes.\n",
      "[18:33:02] Fixed 2 duplicate /gene - greA_1 greA_2\n",
      "[18:33:02] Fixed 6 duplicate /gene - arnC_1 arnC_2 arnC_3 arnC_4 arnC_5 arnC_6\n",
      "[18:33:02] Fixed 2 duplicate /gene - fbp_1 fbp_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - carB_1 carB_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - csoR_1 csoR_2\n",
      "[18:33:02] Fixed 9 duplicate /gene - mshA_1 mshA_2 mshA_3 mshA_4 mshA_5 mshA_6 mshA_7 mshA_8 mshA_9\n",
      "[18:33:02] Fixed 2 duplicate /gene - eno_1 eno_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - mreB_1 mreB_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - sasA_1 sasA_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - xerD_1 xerD_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - atpA_1 atpA_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - ndk_1 ndk_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - argE_1 argE_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - lon1_1 lon1_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - ftsA_1 ftsA_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - atpG_1 atpG_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - spsB_1 spsB_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - pbpG_1 pbpG_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - ruvB_1 ruvB_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - tmk_1 tmk_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - mtgA_1 mtgA_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - mntH_1 mntH_2\n",
      "[18:33:02] Fixed 3 duplicate /gene - prs_1 prs_2 prs_3\n",
      "[18:33:02] Fixed 3 duplicate /gene - rgtE_1 rgtE_2 rgtE_3\n",
      "[18:33:02] Fixed 2 duplicate /gene - polA_1 polA_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - lexA_1 lexA_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - dcd_1 dcd_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - cmk_1 cmk_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - lnrL_1 lnrL_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - rppH_1 rppH_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - nudC_1 nudC_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - fcl_1 fcl_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - rtcB_1 rtcB_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - galE_1 galE_2\n",
      "[18:33:02] Fixed 2 duplicate /gene - idi_1 idi_2\n",
      "[18:33:02] Fixed 4 duplicate /gene - COQ3_1 COQ3_2 COQ3_3 COQ3_4\n",
      "[18:33:02] Fixed 2 duplicate /gene - atpD_1 atpD_2\n",
      "[18:33:02] Fixed 37 colliding /gene names.\n",
      "[18:33:02] Adding /locus_tag identifiers\n",
      "[18:33:02] Assigned 1133 locus_tags to CDS and RNA features.\n",
      "[18:33:02] Writing outputs to ./03_GENOM_ANNOT/bin.10_prokka/\n",
      "[18:33:03] Generating annotation statistics file\n",
      "[18:33:03] Generating Genbank and Sequin files\n",
      "[18:33:03] Running: tbl2asn -V b -a r10k -l paired-ends -M n -N 1 -y 'Annotated using prokka 1.14.6 from https://github.com/tseemann/prokka' -Z \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.err -i \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.fsa 2> /dev/null\n",
      "[tbl2asn-forever] Found Prokka input, correcting dates in .gbf|.sqn files.\n",
      "[tbl2asn-forever] Correcting dates in ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.gbf\n",
      "[tbl2asn-forever] Correcting dates in ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.sqn\n",
      "[tbl2asn-forever] Dates changed from 01-JAN-2019 to 24-AUG-2023\n",
      "[18:33:04] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/errorsummary.val\n",
      "[18:33:04] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.dr\n",
      "[18:33:04] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.fixedproducts\n",
      "[18:33:04] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.ecn\n",
      "[18:33:04] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.val\n",
      "[18:33:04] Repairing broken .GBK output that tbl2asn produces...\n",
      "[18:33:04] Running: sed 's/COORDINATES: profile/COORDINATES:profile/' < \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.gbf > \\.\\/03_GENOM_ANNOT\\/bin\\.10_prokka\\/PROKKA_08242023\\.gbk\n",
      "[18:33:04] Deleting unwanted file: ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.gbf\n",
      "[18:33:04] Output files:\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.err\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.fsa\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.gbk\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.faa\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.gff\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.tbl\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.fna\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.ffn\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.sqn\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.txt\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.tsv\n",
      "[18:33:04] ./03_GENOM_ANNOT/bin.10_prokka/PROKKA_08242023.log\n",
      "[18:33:04] Annotation finished successfully.\n",
      "[18:33:04] Walltime used: 0.50 minutes\n",
      "[18:33:04] If you use this result please cite the Prokka paper:\n",
      "[18:33:04] Seemann T (2014) Prokka: rapid prokaryotic genome annotation. Bioinformatics. 30(14):2068-9.\n",
      "[18:33:04] Type 'prokka --citation' for more details.\n",
      "[18:33:04] Thank you, come again.\n",
      "[18:33:04] This is prokka 1.14.6\n",
      "[18:33:04] Written by Torsten Seemann <torsten.seemann@gmail.com>\n",
      "[18:33:04] Homepage is https://github.com/tseemann/prokka\n",
      "[18:33:04] Local time is Thu Aug 24 18:33:04 2023\n",
      "[18:33:04] You are lu87neb\n",
      "[18:33:04] Operating system is linux\n",
      "[18:33:04] You have BioPerl 1.7.8\n",
      "Argument \"1.7.8\" isn't numeric in numeric lt (<) at /data/mambaforge/envs/genome_annot/bin/prokka line 259.\n",
      "[18:33:04] System has 48 cores.\n",
      "[18:33:04] Will use maximum of 8 cores.\n",
      "[18:33:04] Annotating as >>> Bacteria <<<\n",
      "[18:33:04] Creating new output folder: ./03_GENOM_ANNOT/bin.1_prokka\n",
      "[18:33:04] Running: mkdir -p \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\n",
      "[18:33:04] Using filename prefix: PROKKA_08242023.XXX\n",
      "[18:33:04] Setting HMMER_NCPU=1\n",
      "[18:33:04] Writing log to: ./03_GENOM_ANNOT/bin.1_prokka/PROKKA_08242023.log\n",
      "[18:33:04] Command: /data/mambaforge/envs/genome_annot/bin/prokka --outdir ./03_GENOM_ANNOT/bin.1_prokka --locustag bin.1 ./02_ASSEMBLY_BINNING/bin_refinement_spades_metawrap/metawrap_70_10_bins/bin.1.fa\n",
      "[18:33:04] Appending to PATH: /data/mambaforge/envs/genome_annot/bin\n",
      "[18:33:04] Looking for 'aragorn' - found /data/mambaforge/envs/genome_annot/bin/aragorn\n",
      "[18:33:04] Determined aragorn version is 001002 from 'ARAGORN v1.2.41 Dean Laslett'\n",
      "[18:33:04] Looking for 'barrnap' - found /data/mambaforge/envs/genome_annot/bin/barrnap\n",
      "[18:33:05] Determined barrnap version is 000009 from 'barrnap 0.9'\n",
      "[18:33:05] Looking for 'blastp' - found /data/mambaforge/envs/genome_annot/bin/blastp\n",
      "[18:33:05] Determined blastp version is 002013 from 'blastp: 2.13.0+'\n",
      "[18:33:05] Looking for 'cmpress' - found /data/mambaforge/envs/genome_annot/bin/cmpress\n",
      "[18:33:05] Determined cmpress version is 001001 from '# INFERNAL 1.1.4 (Dec 2020)'\n",
      "[18:33:05] Looking for 'cmscan' - found /data/mambaforge/envs/genome_annot/bin/cmscan\n",
      "[18:33:05] Determined cmscan version is 001001 from '# INFERNAL 1.1.4 (Dec 2020)'\n",
      "[18:33:05] Looking for 'egrep' - found /usr/bin/egrep\n",
      "[18:33:05] Looking for 'find' - found /usr/bin/find\n",
      "[18:33:05] Looking for 'grep' - found /usr/bin/grep\n",
      "[18:33:05] Looking for 'hmmpress' - found /data/mambaforge/envs/genome_annot/bin/hmmpress\n",
      "[18:33:05] Determined hmmpress version is 003003 from '# HMMER 3.3.2 (Nov 2020); http://hmmer.org/'\n",
      "[18:33:05] Looking for 'hmmscan' - found /data/mambaforge/envs/genome_annot/bin/hmmscan\n",
      "[18:33:05] Determined hmmscan version is 003003 from '# HMMER 3.3.2 (Nov 2020); http://hmmer.org/'\n",
      "[18:33:05] Looking for 'java' - found /data/mambaforge/envs/genome_annot/bin/java\n",
      "[18:33:05] Looking for 'makeblastdb' - found /data/mambaforge/envs/genome_annot/bin/makeblastdb\n",
      "[18:33:05] Determined makeblastdb version is 002013 from 'makeblastdb: 2.13.0+'\n",
      "[18:33:05] Looking for 'minced' - found /data/mambaforge/envs/genome_annot/bin/minced\n",
      "[18:33:05] Determined minced version is 004002 from 'minced 0.4.2'\n",
      "[18:33:05] Looking for 'parallel' - found /data/mambaforge/envs/genome_annot/bin/parallel\n",
      "[18:33:05] Determined parallel version is 20220722 from 'GNU parallel 20220722'\n",
      "[18:33:05] Looking for 'prodigal' - found /data/mambaforge/envs/genome_annot/bin/prodigal\n",
      "[18:33:05] Determined prodigal version is 002006 from 'Prodigal V2.6.3: February, 2016'\n",
      "[18:33:05] Looking for 'prokka-genbank_to_fasta_db' - found /data/mambaforge/envs/genome_annot/bin/prokka-genbank_to_fasta_db\n",
      "[18:33:05] Looking for 'sed' - found /usr/bin/sed\n",
      "[18:33:05] Looking for 'tbl2asn' - found /data/mambaforge/envs/genome_annot/bin/tbl2asn\n",
      "[18:33:05] Determined tbl2asn version is 025007 from 'tbl2asn 25.7   arguments:'\n",
      "[18:33:05] Using genetic code table 11.\n",
      "[18:33:05] Loading and checking input file: ./02_ASSEMBLY_BINNING/bin_refinement_spades_metawrap/metawrap_70_10_bins/bin.1.fa\n",
      "[18:33:05] Wrote 148 contigs totalling 4480464 bp.\n",
      "[18:33:05] Predicting tRNAs and tmRNAs\n",
      "[18:33:05] Running: aragorn -l -gc11  -w \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\\/PROKKA_08242023\\.fna\n",
      "[18:33:06] 1 tRNA-Leu [62356,62442] 35 (cag)\n",
      "[18:33:06] 1 tRNA-Asn [64955,65030] 34 (gtt)\n",
      "[18:33:06] 1 tRNA-Leu c[9038,9123] 35 (taa)\n",
      "[18:33:06] 2 tRNA-Cys c[9134,9207] 33 (gca)\n",
      "[18:33:06] 3 tRNA-Gly c[9260,9335] 34 (gcc)\n",
      "[18:33:06] 4 tRNA-Lys c[15784,15859] 34 (ctt)\n",
      "[18:33:06] 1 tRNA-Gly c[38883,38956] 33 (ccc)\n",
      "[18:33:06] 1 tRNA-Gln c[58992,59066] 33 (ctg)\n",
      "[18:33:06] 2 tRNA-Met c[59072,59148] 35 (cat)\n",
      "[18:33:06] 3 tRNA-Gln c[59181,59255] 33 (ttg)\n",
      "[18:33:06] 4 tRNA-Leu c[59285,59369] 35 (tag)\n",
      "[18:33:06] 5 tRNA-Met c[59377,59453] 35 (cat)\n",
      "[18:33:06] 1 tRNA-Leu [57330,57414] 35 (caa)\n",
      "[18:33:06] 1 tRNA-Met [57799,57875] 35 (cat)\n",
      "[18:33:06] 1 tRNA-Phe [18061,18136] 34 (gaa)\n",
      "[18:33:06] 1 tRNA-Phe [47808,47883] 34 (gaa)\n",
      "[18:33:06] 1 tRNA-Pro c[48645,48721] 35 (tgg)\n",
      "[18:33:06] 1 tRNA-Lys c[46855,46930] 34 (ttt)\n",
      "[18:33:06] 2 tRNA-Val c[47130,47205] 34 (tac)\n",
      "[18:33:06] 3 tRNA-Val c[47250,47325] 34 (tac)\n",
      "[18:33:06] 4 tRNA-Val c[47370,47445] 34 (tac)\n",
      "[18:33:06] 1 tRNA-Met c[6783,6859] 35 (cat)\n",
      "[18:33:06] 2 tRNA-Leu c[6920,7006] 35 (gag)\n",
      "[18:33:06] 1 tRNA-SeC [16041,16135] 35 (tca)\n",
      "[18:33:06] 1 tRNA-Thr [64,139] 34 (ggt)\n",
      "[18:33:06] 1 tRNA-Tyr c[36254,36338] 35 (gta)\n",
      "[18:33:06] 1 tRNA-Pro [3440,3516] 35 (ggg)\n",
      "[18:33:06] 2 tRNA-Thr c[35551,35626] 34 (tgt)\n",
      "[18:33:06] 3 tRNA-Ser c[35632,35719] 35 (tga)\n",
      "[18:33:06] 1 tRNA-Ser [31217,31309] 35 (gct)\n",
      "[18:33:06] 2 tRNA-Arg [31318,31394] 35 (acg)\n",
      "[18:33:06] 3 tRNA-Arg [31452,31528] 35 (acg)\n",
      "[18:33:06] 1 tRNA-Gly c[21590,21665] 34 (gcc)\n",
      "[18:33:06] 1 tRNA-Trp c[32614,32689] 34 (cca)\n",
      "[18:33:06] 1 tRNA-Lys c[4889,4964] 34 (ttt)\n",
      "[18:33:06] 2 tRNA-Lys c[4992,5067] 34 (ttt)\n",
      "[18:33:06] 3 tmRNA [32164,32527] 91,126 ANDENYALAA**\n",
      "[18:33:06] 1 tRNA-Val [28847,28923] 35 (gac)\n",
      "[18:33:06] 2 tRNA-Val [28983,29059] 35 (gac)\n",
      "[18:33:06] 1 tRNA-Arg [14363,14437] 34 (cct)\n",
      "[18:33:06] 1 tRNA-Met [12316,12392] 35 (cat)\n",
      "[18:33:06] 2 tRNA-Met [12516,12592] 35 (cat)\n",
      "[18:33:06] 1 tRNA-Ser c[16410,16497] 35 (tga)\n",
      "[18:33:06] 1 tRNA-Asn c[18529,18604] 34 (gtt)\n",
      "[18:33:06] 1 tRNA-Ser c[2787,2876] 35 (gga)\n",
      "[18:33:06] 1 tRNA-Arg [24778,24854] 35 (ccg)\n",
      "[18:33:06] 2 tRNA-His [24938,25013] 34 (gtg)\n",
      "[18:33:06] 3 tRNA-Leu [25034,25120] 35 (cag)\n",
      "[18:33:06] 4 tRNA-Pro [25196,25272] 35 (tgg)\n",
      "[18:33:06] 1 tRNA-Asn [15554,15629] 34 (gtt)\n",
      "[18:33:06] 1 tRNA-Arg c[9345,9421] 35 (tct)\n",
      "[18:33:06] 1 tRNA-Ala [235,310] 34 (ggc)\n",
      "[18:33:06] 2 tRNA-Ala [388,463] 34 (ggc)\n",
      "[18:33:06] 1 tRNA-Asp c[2149,2225] 35 (gtc)\n",
      "[18:33:06] 1 tRNA-Ser [4926,5015] 35 (gga)\n",
      "[18:33:06] Found 55 tRNAs\n",
      "[18:33:06] Predicting Ribosomal RNAs\n",
      "[18:33:06] Running Barrnap with 8 threads\n",
      "[18:33:07] 1 NODE_302_length_37462_cov_18.099046 181 5S ribosomal RNA\n",
      "[18:33:07] Found 1 rRNAs\n",
      "[18:33:07] Skipping ncRNA search, enable with --rfam if desired.\n",
      "[18:33:07] Total of 55 tRNA + rRNA features\n",
      "[18:33:07] Searching for CRISPR repeats\n",
      "[18:33:08] CRISPR1 NODE_164_length_79444_cov_17.168247 22507 with 9 spacers\n",
      "[18:33:08] CRISPR2 NODE_182_length_68565_cov_16.191724 8246 with 6 spacers\n",
      "[18:33:08] CRISPR3 NODE_256_length_48760_cov_16.904301 3971 with 4 spacers\n",
      "[18:33:08] Found 3 CRISPRs\n",
      "[18:33:08] Predicting coding sequences\n",
      "[18:33:08] Contigs total 4480464 bp, so using single mode\n",
      "[18:33:08] Running: prodigal -i \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\\/PROKKA_08242023\\.fna -c -m -g 11 -p single -f sco -q\n",
      "[18:33:13] Excluding CDS which overlaps existing RNA (tRNA) at NODE_238_length_53524_cov_16.378350:18049..18228 on + strand\n",
      "[18:33:13] Excluding CDS which overlaps existing RNA (tRNA) at NODE_243_length_51054_cov_15.971666:48601..48786 on + strand\n",
      "[18:33:13] Excluding CDS which overlaps existing RNA (tRNA) at NODE_308_length_36599_cov_16.265789:36253..36396 on + strand\n",
      "[18:33:13] Excluding CDS which overlaps existing RNA (tRNA) at NODE_311_length_36097_cov_17.260363:3428..3949 on + strand\n",
      "[18:33:13] Excluding CDS which overlaps existing RNA (tRNA) at NODE_324_length_33603_cov_18.555234:21579..21773 on + strand\n",
      "[18:33:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_386_length_26042_cov_16.088621:25083..25418 on - strand\n",
      "[18:33:14] Excluding CDS which overlaps existing RNA (tRNA) at NODE_558_length_10452_cov_16.279600:199..363 on + strand\n",
      "[18:33:14] Found 3939 CDS\n",
      "[18:33:14] Connecting features back to sequences\n",
      "[18:33:14] Not using genus-specific database. Try --usegenus to enable it.\n",
      "[18:33:14] Annotating CDS, please be patient.\n",
      "[18:33:14] Will use 8 CPUs for similarity searching.\n",
      "[18:33:15] There are still 3939 unannotated CDS left (started with 3939)\n",
      "[18:33:15] Will use blast to search against /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/IS with 8 CPUs\n",
      "[18:33:15] Running: cat \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\\/PROKKA_08242023\\.IS\\.tmp\\.1136146\\.faa | parallel --gnu --plain -j 8 --block 79048 --recstart '>' --pipe blastp -query - -db /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/IS -evalue 1e-30 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\\/PROKKA_08242023\\.IS\\.tmp\\.1136146\\.blast 2> /dev/null\n",
      "[18:33:25] Deleting unwanted file: ./03_GENOM_ANNOT/bin.1_prokka/PROKKA_08242023.IS.tmp.1136146.faa\n",
      "[18:33:25] Deleting unwanted file: ./03_GENOM_ANNOT/bin.1_prokka/PROKKA_08242023.IS.tmp.1136146.blast\n",
      "[18:33:26] There are still 3919 unannotated CDS left (started with 3939)\n",
      "[18:33:26] Will use blast to search against /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/AMR with 8 CPUs\n",
      "[18:33:26] Running: cat \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\\/PROKKA_08242023\\.AMR\\.tmp\\.1136146\\.faa | parallel --gnu --plain -j 8 --block 78862 --recstart '>' --pipe blastp -query - -db /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/AMR -evalue 1e-300 -qcov_hsp_perc 90 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\\/PROKKA_08242023\\.AMR\\.tmp\\.1136146\\.blast 2> /dev/null\n",
      "[18:33:34] Deleting unwanted file: ./03_GENOM_ANNOT/bin.1_prokka/PROKKA_08242023.AMR.tmp.1136146.faa\n",
      "[18:33:34] Deleting unwanted file: ./03_GENOM_ANNOT/bin.1_prokka/PROKKA_08242023.AMR.tmp.1136146.blast\n",
      "[18:33:35] There are still 3919 unannotated CDS left (started with 3939)\n",
      "[18:33:35] Will use blast to search against /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/sprot with 8 CPUs\n",
      "[18:33:35] Running: cat \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\\/PROKKA_08242023\\.sprot\\.tmp\\.1136146\\.faa | parallel --gnu --plain -j 8 --block 78862 --recstart '>' --pipe blastp -query - -db /data/mambaforge/envs/genome_annot/db/kingdom/Bacteria/sprot -evalue 1e-09 -qcov_hsp_perc 80 -num_threads 1 -num_descriptions 1 -num_alignments 1 -seg no > \\.\\/03_GENOM_ANNOT\\/bin\\.1_prokka\\/PROKKA_08242023\\.sprot\\.tmp\\.1136146\\.blast 2> /dev/null\n"
     ]
    }
   ],
   "source": [
    "### Let's annotate the binned genomes\n",
    "cd ~/data/MBGW223\n",
    "#mkdir 03_GENOM_ANNOT\n",
    "conda deactivate\n",
    "conda activate genome_annot\n",
    "for file in ./02_ASSEMBLY_BINNING/bin_refinement_spades_metawrap/metawrap_70_10_bins/*.fa; do filename=$(basename -- \"$file\"); prefix=\"${filename%.*}\"; prokka --outdir ./03_GENOM_ANNOT/\"$prefix\"_prokka --locustag $prefix $file; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c89da4e5-82e8-41b4-94d5-1736cd1d6a61",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>QUESTION/TASK:</b> \n",
    "Have a look at the prokka output folder to get an idea about what kind of output files are generated.\n",
    "<ul>\n",
    "  <li>What is the difference between .gff and .gbk files?</li>\n",
    "  <li>Dive into the nucleic acid fasta (.ffn), can you find any 16S rRNA genes?</li>\n",
    "</ul>\n",
    "</div>\n",
    "\n",
    "### Probing potentially present 16S rRNA genes\n",
    "\n",
    "If yes, download them and browse to NCBI blast select nucleotide blast, upload the sequences and choose 16S ribosomal RNA sequences (Bacteria/Archaea) as reference. Besides from using NCBI's BLAST suite you can also make use of resources for rRNA sequence analysis, as for instance SILVA. Browse to SILVA, upload your sequences, tick “Search and Classify” and run the aligner. Does the 16S rRNA gene-based taxonomy match the taxonomy determined by `GTDB-Tk`?\n",
    "\n",
    "### Assessing the metabolic potential | KEGG pathway mapping\n",
    "\n",
    "Let's (superficially) check the metabolic potential of our bins. In principle, this is possible by taking our genome annotations and checking which functions are encoded. This is extremely laborious. If you are interested in genes linked to a particular pathway/function (e.g. autotrophy) you would be forced to check the annotation for all the enzymatic functions involved in that. \n",
    "\n",
    "In order to bypass this issue, we take advantage of [`KEGG`](https://www.genome.jp/kegg/), the Kyoto Encyclopedia of Genomes and Genes. Among many other things, KEGG contains a sophisticated collection of pathway maps, which can be used for metabolic reconstructions. Each enzymatic function of a particular pathway has an asssigned K-number. If you have a list of K-numbers you can map them on the KEGG pathway maps as a whole and check which pathways they are linked to. \n",
    "\n",
    "This is exactly what we want to do now. In order to get lists of K-numbers for our bins we have to run database searches for all encoded amino acid sequences against the KEGG database.\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>QUESTION/TASK:</b> \n",
    "<ul>\n",
    "  <li>Move all .faa files from the prokka output folders into one folder.</li>\n",
    "  <li>Concatenate them cat *.faa > bins_AAseqs.faa</li>\n",
    "  <li>Download this concatenated file and upload it to [`BlastKOALA`](https://www.kegg.jp/blastkoala/)</li>\n",
    "</ul>\n",
    "</div>\n",
    "\n",
    "![BLASTKOALA mask](img/blastkoala.png)\n",
    "<font size=\"2\"> The BLASTKOALA job submission mask. </font>\n",
    "\n",
    "* Check your mails, confirm the BlastKOALA job and wait for the results\n",
    "* Once the job is done you can download the results (a tab-separated table, gene ids \\t K-number) and upload them to [`KEGG Mapper`](https://www.genome.jp/kegg/tool/map_pathway.html).\n",
    "\n",
    "![BLASTKOALA mask](img/keggmapper.png)\n",
    "\n",
    "<font size=\"2\"> The KEGG MAPPER job submission mask. </font>\n",
    "\n",
    "<div class=\"alert alert-block alert-warning\">\n",
    "<b>QUESTION/TASK:</b> \n",
    "Go through all bins and take a look at selected metabolic aspects:\n",
    "<ul>\n",
    "  <li>Do all genome bins encode the complete set of genes for glycolysis? » Check “Carbohydrate metabolism”</li>\n",
    "  <li>Do some bins have a broader substrate utilization potential? » Check “Carbohydrate metabolism”</li>\n",
    "  <li>Do all bins encode peptidoglycan biosynthesis? » Check “Glycan biosynthesis”</li>\n",
    "  <li>Do you find pathways of biogeochemical relevance? » Check “Methane/Nitrogen metabolism”</li>\n",
    "  <li>Is there any bin with an extended potential for secondary metabolite biosynthesis? » Check \"Type I polyketide structures\"</li>\n",
    "  <li>Does one of the bins feature eukaryotic signature proteins? » Check \"Protein export\" & \"Viral life cycle - HIV1\"</li>\n",
    "</ul>\n",
    "</div>\n",
    "\n",
    "<sub> © Carl-Eric Wegner, 2023-08 </sub>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bf53d592-d37c-4dd2-9d05-d7a6b4cb307d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
