{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "4f57f838-1b57-4225-9c70-4983c4d57c37",
   "metadata": {},
   "source": [
    "# ChemBioSys and AquaDiva | Genome-resolved metagenomics workshop\n",
    "## Functional genome mining - finding biosynthetic gene clusters associated with specialized metabolites\n",
    "\n",
    "In this section of the workshop, we will provide an overview of biosynthetic gene clusters, the functions they code for in genomes, and explore some tools used to find them."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a9521502-381d-485a-9acc-41622f70ca04",
   "metadata": {},
   "source": [
    "## So what are specialized metabolites and biosynthetic gene clusters?\n",
    "\n",
    "Specialized metabolites, also often referred to as secondary metabolites (although not all secondary metabolites are so specialized) include antibacterial, antifungal, anticancer compounds, often produced by bacteria and fungi. These compounds represent many different compound classes, of which non-ribosomal peptide synthases (NRPS), polyketide synthases (PKS), ribosomally synthesised and post-translationally modified peptides (RiPP), terpenes, bacteriocins are most well known. These compounds have historically been discovered through lab-based bioassays combined with liquid chromatography, mass spectometry and nuclear magnetic resonance (NMR) spectroscopy. \n",
    "\n",
    "Along with high throughput sequencing becoming a cost-effective and widely used method in molecular biology, new avenues for discovery of specialized metabolites, and also studying the evolution and distribution of genetic content associated with these metabolites have been developed. \n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9522f3e-0b3b-4b8f-922d-70306aa3d061",
   "metadata": {},
   "source": [
    "## Exercise - Explore a biosynthetic gene cluster.\n",
    "\n",
    "[Open the Mibig database here](https://mibig.secondarymetabolites.org/repository)\n",
    "\n",
    "Search the repository for an antibiotic - perhaps one that you have taken in the past, or one that interests you.\n",
    "\n",
    "Look through the information provided, like the length of the BGC and the number and type of genes. If you scroll down you will find some more information, such as the organism or genome where it was recorded. \n",
    "\n",
    "The menu will take you through the compound structure, NRPS or PKS modules and accession numbers for individual genes.\n",
    "\n",
    "How many genes are present in the BGC you selected?\n",
    "How long is the cluster?\n",
    "Do you find some regulatory genes, transporter genes and genes of unknown function, along with the core biosynthesis genes?\n",
    "What organisms produce it?\n",
    "Is it an NRPS or PKS or something else?\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64578868-71f9-483f-8891-ef10bbf9e519",
   "metadata": {},
   "source": [
    "## Using antiSMASH to detect biosynthetic gene clusters in metagenome assembled genomes\n",
    "\n",
    "We will use antiSMASH (antibiotic and Secondary Metabolite Analysis Server), which is currently the premier tool for detection of biosynthetic gene clusters.\n",
    "\n",
    "antiSMASH will search all genes in your genome against a set of core BGC-related genes in its database. [open this link for more infomation.](https://docs.antismash.secondarymetabolites.org/understanding_output/regions/) Upon detection any of these genes, it will then use a set of \"rules\" for each type of BGC it detects. These rules are essentially a set of HMM for each rule and upon detection, the cluster gets identified according to the rule. \n",
    "\n",
    "For example, in PKS gene clusters contain regions called ketosynthase (KS) and acyltransferase (AT) domains which are conserved to some degree. For this reason it is possible to build HMMs to detect them. The same is true for NRPS clusters, which have condensation (CDS) and A- (adenylation) domains which enable its detection.\n",
    "\n",
    "You can see all of the antiSMASH rules [here](https://docs.antismash.secondarymetabolites.org/glossary/#clustertypes). If you are interested in seeing more of the inner workings, open the rule files in the [antiSMASH source.](https://github.com/antismash/antismash/blob/master/antismash/detection/hmm_detection/cluster_rules/strict.txt)\n",
    " \n",
    "[Now open the antiSMASH webserver.](https://antismash.secondarymetabolites.org/#!/start)\n",
    "\n",
    "[And lets look at a typical set of results.](https://antismash.secondarymetabolites.org/upload/bacteria-75a3865d-f2be-4952-bf50-07897c40434f/index.html#r1c1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49c5d564-8386-40e8-b4aa-363a20e04551",
   "metadata": {},
   "source": [
    "## antiSMASH input options\n",
    "\n",
    "### KnownClusterBlast\r\n",
    "\r\n",
    "The identified clusters are searched against the MIBiG repository. MIBiG is a hand curated data collection of biosynthetic gene clusters, which have been experimentally characterized.### \r\n",
    "\r\n",
    "ClusterBlast\r\n",
    "\r\n",
    "The identified clusters are searched against a comprehensive gene cluster database and similar clusters are identified. The algorithm used here is inspired by MultiGeneBlast. It runs BlastP using each amino acid sequence from a detected gene cluster as a query on a large database of predicted protein sequences from secondary metabolite biosynthetic gene clusters, and pools the results to identify the gene clusters that are most homologous to the gene cluster that was detected in your query nucleotide sequence. Please note that selecting this option increases the runtime significan### tly.\r\n",
    "\r\n",
    "SubClusterBlast\r\n",
    "\r\n",
    "The identified clusters are searched against a database containing operons involved in the biosynthesis of common secondary metabolite building blocks (e.g. the biosynthesis of non-proteinogenic amino###  acids).\r\n",
    "\r\n",
    "ActiveSiteFinder\r\n",
    "\r\n",
    "Active sites of several highly conserved biosynthetic enzymes are detected and variations of the active sites a### re reported.\r\n",
    "\r\n",
    "Cluster Pfam analysis\r\n",
    "\r\n",
    "Each gene product encoded in the detected BGCs is analyzed against the PFAM database. Hits are annotated in the final Genbank/EMBL files that can be downloaded after the analysis is finished. Please note that these results are not displayed on the antiSMASH HTML results page but they are present in the results genbank file that can be downloaded. Also, selecting this option normally increa### ses the runtime.\r\n",
    "\r\n",
    "Pfam-based GO term annotation\r\n",
    "\r\n",
    "This is annotating the Cluster Pfam analysis described above with GO term annotations. Please note that these results are not displayed on the antiSMASH HTML results page but they are present in the results genbank file that can be downloaded (see \"Understanding the output\" section of the documentation for instructions on how to download the results). Also, selecting this option normally increases the runtime"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d58bc03f-b06b-4191-a16c-8be0e3e4c677",
   "metadata": {},
   "source": [
    "## Running antiSMASH standalone\n",
    "\n",
    "Submitting your metagenome bins individually to the antiSMASH webserver will be incredibly time-consuming and counterproductive. Instead, we use the antiSMASH standalone version. This also provides the opportunity for further customizing your analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "93e1a1c4-b759-408d-a568-d122b24b4714",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/home/qe63rov/data/programs/conda/antismash_latest) \n",
      "(/home/qe63rov/data/programs/conda/antismash_latest) \n",
      "\n",
      "########### antiSMASH 7.0.1 #############\n",
      "\n",
      "usage: antismash [-h] [--help-showall] [-t {bacteria,fungi}] [-c CPUS]\n",
      "                 [--databases PATH] [--output-dir OUTPUT_DIR]\n",
      "                 [--output-basename OUTPUT_BASENAME] [--reuse-results PATH]\n",
      "                 [--limit LIMIT]\n",
      "                 [--abort-on-invalid-records | --no-abort-on-invalid-records]\n",
      "                 [--minlength MINLENGTH] [--start START] [--end END]\n",
      "                 [--write-config-file PATH] [--without-fimo]\n",
      "                 [--executable-paths EXECUTABLE=PATH,EXECUTABLE2=PATH2,...]\n",
      "                 [--allow-long-headers | --no-allow-long-headers] [-v] [-d]\n",
      "                 [--logfile PATH] [--list-plugins] [--check-prereqs]\n",
      "                 [--limit-to-record RECORD_ID] [-V] [--profiling]\n",
      "                 [--skip-sanitisation] [--skip-zip-file] [--minimal]\n",
      "                 [--enable-genefunctions] [--enable-lanthipeptides]\n",
      "                 [--enable-lassopeptides] [--enable-nrps-pks]\n",
      "                 [--enable-sactipeptides] [--enable-t2pks]\n",
      "                 [--enable-thiopeptides] [--enable-tta] [--enable-html]\n",
      "                 [--fullhmmer]\n",
      "                 [--fullhmmer-pfamdb-version FULLHMMER_PFAMDB_VERSION]\n",
      "                 [--hmmdetection-strictness {strict,relaxed,loose}]\n",
      "                 [--hmmdetection-fungal-cutoff-multiplier HMMDETECTION_FUNGAL_CUTOFF_MULTIPLIER]\n",
      "                 [--hmmdetection-fungal-neighbourhood-multiplier HMMDETECTION_FUNGAL_NEIGHBOURHOOD_MULTIPLIER]\n",
      "                 [--sideload JSON] [--sideload-simple ACCESSION:START-END]\n",
      "                 [--sideload-by-cds LOCUS1,LOCUS2,...]\n",
      "                 [--sideload-size-by-cds NUCLEOTIDES] [--cassis]\n",
      "                 [--clusterhmmer]\n",
      "                 [--clusterhmmer-pfamdb-version CLUSTERHMMER_PFAMDB_VERSION]\n",
      "                 [--tigrfam] [--asf] [--cc-mibig]\n",
      "                 [--cc-custom-dbs FILE1,FILE2,...] [--cb-general]\n",
      "                 [--cb-subclusters] [--cb-knownclusters]\n",
      "                 [--cb-nclusters count] [--cb-min-homology-scale LIMIT]\n",
      "                 [--pfam2go] [--rre] [--rre-cutoff RRE_CUTOFF]\n",
      "                 [--rre-minlength RRE_MIN_LENGTH] [--smcog-trees] [--tfbs]\n",
      "                 [--tfbs-pvalue TFBS_PVALUE] [--tfbs-range TFBS_RANGE]\n",
      "                 [--tta-threshold TTA_THRESHOLD] [--html-title HTML_TITLE]\n",
      "                 [--html-description HTML_DESCRIPTION] [--html-start-compact]\n",
      "                 [--genefinding-tool {glimmerhmm,prodigal,prodigal-m,none,error}]\n",
      "                 [--genefinding-gff3 GFF3_FILE]\n",
      "                 [SEQUENCE ...]\n",
      "\n",
      "\n",
      "arguments:\n",
      "  SEQUENCE  GenBank/EMBL/FASTA file(s) containing DNA.\n",
      "\n",
      "--------\n",
      "Options\n",
      "--------\n",
      "options:\n",
      "\n",
      "Help options:\n",
      "\n",
      "  -h, --help            Show basic help text.\n",
      "  --help-showall        Show full list of arguments.\n",
      "\n",
      "Basic analysis options:\n",
      "\n",
      "  -t {bacteria,fungi}, --taxon {bacteria,fungi}\n",
      "                        Taxonomic classification of input sequence. (default:\n",
      "                        bacteria)\n",
      "  -c CPUS, --cpus CPUS  How many CPUs to use in parallel. (default for this\n",
      "                        machine: 48)\n",
      "  --databases PATH      Root directory of the databases (default: /home/qe63ro\n",
      "                        v/data/programs/conda/antismash_latest/lib/python3.10/\n",
      "                        site-packages/antismash/databases).\n",
      "\n",
      "Output options:\n",
      "\n",
      "  --output-dir OUTPUT_DIR\n",
      "                        Directory to write results to.\n",
      "  --output-basename OUTPUT_BASENAME\n",
      "                        Base filename to use for output files within the\n",
      "                        output directory.\n",
      "  --html-title HTML_TITLE\n",
      "                        Custom title for the HTML output page (default is\n",
      "                        input filename).\n",
      "  --html-description HTML_DESCRIPTION\n",
      "                        Custom description to add to the output.\n",
      "  --html-start-compact  Use compact view by default for overview page.\n",
      "\n",
      "Additional analysis:\n",
      "\n",
      "  --fullhmmer           Run a whole-genome HMMer analysis.\n",
      "  --cassis              Motif based prediction of SM gene cluster regions.\n",
      "  --clusterhmmer        Run a cluster-limited HMMer analysis.\n",
      "  --tigrfam             Annotate clusters using TIGRFam profiles.\n",
      "  --asf                 Run active site finder analysis.\n",
      "  --cc-mibig            Run a comparison against the MIBiG dataset\n",
      "  --cb-general          Compare identified clusters against a database of\n",
      "                        antiSMASH-predicted clusters.\n",
      "  --cb-subclusters      Compare identified clusters against known subclusters\n",
      "                        responsible for synthesising precursors.\n",
      "  --cb-knownclusters    Compare identified clusters against known gene\n",
      "                        clusters from the MIBiG database.\n",
      "  --pfam2go             Run Pfam to Gene Ontology mapping module.\n",
      "  --rre                 Run RREFinder precision mode on all RiPP gene\n",
      "                        clusters.\n",
      "  --smcog-trees         Generate phylogenetic trees of sec. met. cluster\n",
      "                        orthologous groups.\n",
      "  --tfbs                Run TFBS finder on all gene clusters.\n",
      "  --tta-threshold TTA_THRESHOLD\n",
      "                        Lowest GC content to annotate TTA codons at (default:\n",
      "                        0.65).\n",
      "\n",
      "Advanced options:\n",
      "\n",
      "  --reuse-results PATH  Use the previous results from the specified json\n",
      "                        datafile\n",
      "  --limit LIMIT         Only process the largest <limit> records (default:\n",
      "                        -1). -1 to disable\n",
      "  --abort-on-invalid-records, --no-abort-on-invalid-records\n",
      "                        Abort runs when encountering invalid records instead\n",
      "                        of skipping them (default: True)\n",
      "  --minlength MINLENGTH\n",
      "                        Only process sequences larger than <minlength>\n",
      "                        (default: 1000).\n",
      "  --start START         Start analysis at nucleotide specified.\n",
      "  --end END             End analysis at nucleotide specified\n",
      "  --write-config-file PATH\n",
      "                        Write a config file to the supplied path\n",
      "  --without-fimo        Run without FIMO (lowers accuracy of RiPP precursor\n",
      "                        predictions)\n",
      "  --executable-paths EXECUTABLE=PATH,EXECUTABLE2=PATH2,...\n",
      "                        A comma separated list of executable name->path pairs\n",
      "                        to override any on the system path.E.g.\n",
      "                        diamond=/alternate/path/to/diamond,hmmpfam2=hmm2pfam\n",
      "  --allow-long-headers, --no-allow-long-headers\n",
      "                        Should sequence identifiers longer than 16 characters\n",
      "                        be allowed (default: True)\n",
      "\n",
      "Debugging & Logging options:\n",
      "\n",
      "  -v, --verbose         Print verbose status information to stderr.\n",
      "  -d, --debug           Print debugging information to stderr.\n",
      "  --logfile PATH        Also write logging output to a file.\n",
      "  --list-plugins        List all available sec. met. detection modules.\n",
      "  --check-prereqs, --prepare-data\n",
      "                        Check if all prerequisites are met, preparing data\n",
      "                        files where possible.\n",
      "  --limit-to-record RECORD_ID\n",
      "                        Limit analysis to the record with ID record_id\n",
      "  -V, --version         Display the version number and exit.\n",
      "  --profiling           Generate a profiling report, disables multiprocess\n",
      "                        python.\n",
      "  --skip-sanitisation   Skip input record sanitisation. Use with care.\n",
      "  --skip-zip-file       Do not create a zip of the output\n",
      "\n",
      "Debugging options for cluster-specific analyses:\n",
      "\n",
      "  --minimal             Only run core detection modules, no analysis modules\n",
      "                        unless explicitly enabled\n",
      "  --enable-genefunctions\n",
      "                        Enable Gene function annotations (default: enabled,\n",
      "                        unless --minimal is specified)\n",
      "  --enable-lanthipeptides\n",
      "                        Enable Lanthipeptides (default: enabled, unless\n",
      "                        --minimal is specified)\n",
      "  --enable-lassopeptides\n",
      "                        Enable lassopeptide precursor prediction (default:\n",
      "                        enabled, unless --minimal is specified)\n",
      "  --enable-nrps-pks     Enable NRPS/PKS analysis (default: enabled, unless\n",
      "                        --minimal is specified)\n",
      "  --enable-sactipeptides\n",
      "                        Enable sactipeptide detection (default: enabled,\n",
      "                        unless --minimal is specified)\n",
      "  --enable-t2pks        Enable type II PKS analysis (default: enabled, unless\n",
      "                        --minimal is specified)\n",
      "  --enable-thiopeptides\n",
      "                        Enable Thiopeptides (default: enabled, unless\n",
      "                        --minimal is specified)\n",
      "  --enable-tta          Enable TTA detection (default: enabled, unless\n",
      "                        --minimal is specified)\n",
      "  --enable-html         Enable HTML output (default: enabled, unless --minimal\n",
      "                        is specified)\n",
      "\n",
      "Full HMMer options:\n",
      "\n",
      "  --fullhmmer-pfamdb-version FULLHMMER_PFAMDB_VERSION\n",
      "                        PFAM database version number (e.g. 27.0) (default:\n",
      "                        latest).\n",
      "\n",
      "HMM detection options:\n",
      "\n",
      "  --hmmdetection-strictness {strict,relaxed,loose}\n",
      "                        Defines which level of strictness to use for HMM-based\n",
      "                        cluster detection, (default: relaxed).\n",
      "  --hmmdetection-fungal-cutoff-multiplier HMMDETECTION_FUNGAL_CUTOFF_MULTIPLIER\n",
      "                        Sets the multiplier for rule cutoffs in fungal inputs\n",
      "                        (default: 1.0).\n",
      "  --hmmdetection-fungal-neighbourhood-multiplier HMMDETECTION_FUNGAL_NEIGHBOURHOOD_MULTIPLIER\n",
      "                        Sets the multiplier for rule neighbourhoods in fungal\n",
      "                        inputs (default: 1.5).\n",
      "\n",
      "Sideload options:\n",
      "\n",
      "  --sideload JSON       Sideload annotations from the JSON file in the given\n",
      "                        paths. Multiple files can be provided, separated by a\n",
      "                        comma.\n",
      "  --sideload-simple ACCESSION:START-END\n",
      "                        Sideload a single subregion in record ACCESSION from\n",
      "                        START to END. Positions are expected to be 0-indexed,\n",
      "                        with START inclusive and END exclusive.\n",
      "  --sideload-by-cds LOCUS1,LOCUS2,...\n",
      "                        Sideload a subregion around each CDS with the given\n",
      "                        locus tags.\n",
      "  --sideload-size-by-cds NUCLEOTIDES\n",
      "                        Additional padding, in nucleotides, of subregions to\n",
      "                        create for sideloaded subregions by CDS. (default:\n",
      "                        20000)\n",
      "\n",
      "Cluster HMMer options:\n",
      "\n",
      "  --clusterhmmer-pfamdb-version CLUSTERHMMER_PFAMDB_VERSION\n",
      "                        PFAM database version number (e.g. 27.0) (default:\n",
      "                        latest).\n",
      "\n",
      "TIGRFam options:\n",
      "\n",
      "ClusterCompare options:\n",
      "\n",
      "  --cc-custom-dbs FILE1,FILE2,...\n",
      "                        A comma separated list of database config files to run\n",
      "                        with\n",
      "\n",
      "ClusterBlast options:\n",
      "\n",
      "  --cb-nclusters count  Number of clusters from ClusterBlast to display,\n",
      "                        cannot be greater than 50. (default: 10)\n",
      "  --cb-min-homology-scale LIMIT\n",
      "                        A minimum scaling factor for the query BGC in\n",
      "                        ClusterBlast results. Valid range: 0.0 - 1.0. Warning:\n",
      "                        some homologous genes may no longer be visible!\n",
      "                        (default: 0.0)\n",
      "\n",
      "NRPS/PKS options:\n",
      "\n",
      "RREfinder options:\n",
      "\n",
      "  --rre-cutoff RRE_CUTOFF\n",
      "                        Bitscore cutoff for RRE pHMM detection (default:\n",
      "                        25.0).\n",
      "  --rre-minlength RRE_MIN_LENGTH\n",
      "                        Minimum amino acid length of RRE domains (default:\n",
      "                        50).\n",
      "\n",
      "Transcription Factor Binding Site options:\n",
      "\n",
      "  --tfbs-pvalue TFBS_PVALUE\n",
      "                        P-value for TFBS threshold setting (default: 1e-05).\n",
      "  --tfbs-range TFBS_RANGE\n",
      "                        The allowable overlap with gene start positions for\n",
      "                        TFBSs in coding regions (default: 50).\n",
      "\n",
      "Gene finding options (ignored when ORFs are annotated):\n",
      "\n",
      "  --genefinding-tool {glimmerhmm,prodigal,prodigal-m,none,error}\n",
      "                        Specify algorithm used for gene finding: GlimmerHMM,\n",
      "                        Prodigal, Prodigal Metagenomic/Anonymous mode, or\n",
      "                        none. The 'error' option will raise an error if\n",
      "                        genefinding is attempted. The 'none' option will not\n",
      "                        run genefinding. (default: error).\n",
      "  --genefinding-gff3 GFF3_FILE\n",
      "                        Specify GFF3 file to extract features from.\n",
      "\n",
      "(/home/qe63rov/data/programs/conda/antismash_latest) \n"
     ]
    }
   ],
   "source": [
    "cd ~/data/workshop/09_ANTISMASH\n",
    "# activate the antismash conda environment\n",
    "conda activate antismash\n",
    "#\n",
    "# View the antiSMASH input options\n",
    "antismash -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79962618-d23a-4984-964d-eda51acae3f5",
   "metadata": {},
   "source": [
    "## Running antiSMASH - important input parameters\n",
    "\n",
    "As you can you, there are many input options. Most of these depend on the purpose of your analysis, like if you are interested in possibly identifying a BGC using known references (less likely in metagenomics), counting the number of biosynthetic gene clusters and getting to most hits possible, or doing further functional exploration and looking for subclusters and other interesting features. \n",
    "\n",
    "If your are not providing an already annotated genbank (.gbk) file, you also need to specify a genefinding tool using the --genefinding-tool option. In our case, we will use \"--genefinding-tool prodigal-m\", which refers to well-known gene prediction tool \"prodigal\" in metagenomic mode and because our input is unannotated fasta files from your binning steps. \n",
    "\n",
    "You could also provide your own gene calling information by providing the gff files along with your input fasta files.\n",
    "\n",
    "Other considerations:\n",
    "\n",
    "- Inluding whole genome HMMer\n",
    "- using a --minlength cutoff\n",
    "- cluster comparison options (results to be interpreted with caution)\n",
    "\n",
    "Thus, the command to run antiSMASH looks like this:\n",
    "\n",
    "### antismash input_file --output-dir antismash_result/bin_name --clusterhmmer --cc-mibig --cb-general --cb-knownclusters --pfam2go --rre --minlenth 7500 --genefinding-tool prodigal-m\n",
    "\n",
    "However, since we have many genomes, we will run antiSMASH using a for-loop. Essentially, this means that we execute antiSMASH on each of our fasta files (from your genome bins).\n",
    "\n",
    "You might see the warning about long sequence headers - you can ignore this, but keep in mind your contig names are now different. Version 7 of antiSMASH allows disabling this function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aa468d81-9abc-4611-9884-98f7ed0eca86",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(/home/qe63rov/data/programs/conda/antismash_latest) \n",
      "(/home/qe63rov/data/programs/conda/antismash_latest) \n",
      "usage: antismash [-h] [options ..] sequence\n",
      "antismash: error: Input file does not exist: ./input/*.fa\n",
      "(/home/qe63rov/data/programs/conda/antismash_latest) \n"
     ]
    },
    {
     "ename": "",
     "evalue": "2",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "# make a directory from where your antismash output will be stored\n",
    "mkdir antismash_result\n",
    "for fasta_file in ./input/*.fa; do base=$(basename \"$fasta_file\" .fa); antismash \"$fasta_file\" --output-dir antismash_result/$base --clusterhmmer --cc-mibig --minlength 7500 --cb-general --cb-knownclusters --pfam2go --rre --genefinding-tool prodigal-m ; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2f64d54-8929-4cf5-97c3-4b7849080f6e",
   "metadata": {},
   "source": [
    "### Lets explore the antiSMASH output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "23afaa9f-4adb-435a-9ddd-573d2e9155f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd antismash_result\n",
    "ls\n",
    "cd Bin_3\n",
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8312c7df-5af7-4aca-b9aa-bfc086d7f60d",
   "metadata": {},
   "source": [
    "The way we specified that antiSMASH uses an output directory with the same name as our genome bins means that we now have a directory with the results for each.\n",
    "\n",
    "In the output, you will see the following:\n",
    "\n",
    "Bin_3.gbk                         - this file contains ORFs identified and annotated according to the information available \n",
    "Bin_3.json                        - if you are good at Python you will want to explore the treasure trove of information stored in antiSMASH json files.\n",
    "NZ_CP009530.1.region001.gbk       - these are the files that contain most details about the clusters, and are most important for most of the downstream processing.\n",
    "NZ_CP009530.1.region002.gbk\n",
    "NZ_CP009530.1.region003.gbk\n",
    "index.html                        - opening this file will present results exactly as we saw it earlier on the antiSMASH webserver. This could be very useful if you have few genomes and target specific BGCs\n",
    "knownclusterblast.txt\n",
    "clusterblast.txt\n",
    "\n",
    "### Next step would be to tabulate the antiSMASH output to really see what we got. To do this we will use two Python scripts borrowed from Zacchary Reitz (https://github.com/zreitz/multismash). One will create a table with BGC counts per genome, and the other will provide more information on specific BGCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef677f45-04f7-41fa-af3f-3dbf1f4c52e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "head -n 30 NZ_CP009530.1.region001.gbk"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b9bb956-988f-41e6-88c5-313313132200",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ../..\n",
    "python count_regions.py example_output/ BGC_counts.tsv\n",
    "python tabulate_regions.py example_output/ BGCs_info.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d57d7d4-0285-4ab1-985c-ed4dc4e89e31",
   "metadata": {},
   "outputs": [],
   "source": [
    "head -n 60 BGC_counts.tsv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "76c06dda-43b7-43a4-9650-fb7eb878dd05",
   "metadata": {},
   "outputs": [],
   "source": [
    "head -n 60 BGC_tab.tsv"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff09c675-5ac9-4a0c-ab3e-730b2be138a9",
   "metadata": {},
   "source": [
    "You can open the two files generated above and see what your output looks like?\n",
    "\n",
    "How many clusters were detected?\n",
    "Which type of BGCs were most common?\n",
    "Which genome had the most BGCs?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9e17e952-e334-497e-8555-6e62a16d39df",
   "metadata": {},
   "outputs": [],
   "source": [
    "wget https://raw.githubusercontent.com/Zander22/metallophore/main/plot_BGCs.R\n",
    "conda deactivate\n",
    "conda activate r\n",
    "Rscript ./plot_BGCs.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bddcd34-24ca-47e8-ba9a-4b3c66b5b701",
   "metadata": {},
   "source": [
    "Well done, you have now completed your first dive into antiSMASH, explored its capabilities and its output.\n",
    "\n",
    "If your main goal is mining metagenomes or metagenome bins to identify as many BGCs as possible, you might be interested to combine different approaches. \n",
    "\n",
    "### Some other BGC identification tools include:\n",
    "[GECCO](https://gecco.embl.de/) alternative, de novo approach to find BGCs\r",
    "[\n",
    "DeepBG](https://github.com/Merck/deepbgc)C deep learning approach for BGC detection[\r\n",
    "biosyntheticSPAd](https://cab.spbu.ru/software/biosyntheticspades/)es reconstructs BGCs in assembly graphs from genomic and metagenomics data set\n",
    "### s\r\n",
    "Class-specific BGC detecti[on\r\n",
    "RiPP R](https://ripp.rodeo/index.html)ODEO RiPP discovery and genomic cont[ext\r\n",
    "B](https://github.com/annejong/BAGEL4)AGEL4 search DNA for bacteriocins and R[iPPs\r\n",
    "De](http://deepripp.magarveylab.ca/)epRipp neural network-based detection of [RiPPs\r\n",
    "dec](https://github.com/Alexamk/decRiPPter)RiPPter designed to detect novel RiPP BGCs not restricted to specific genetic m[arkers\r\n",
    "t](https://github.com/chevrm/transPACT)ransPACT trans-acetyltransferase polyketide synthase d\n",
    "\n",
    "If you are interested in any of these tools, perhaps you can try them out in th BYOD session.etection"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d17a1c7-55fe-4b07-b509-9cdb049ebbd8",
   "metadata": {},
   "source": [
    "## Clustering BGCs\n",
    "\n",
    "There are many tools available which allows the clustering of BGCs. There are many reasons to do this, such as aiding in identification, doing evolutionary mining analysis or assessing whether BGCs are potentially divergent from anything knwon.\n",
    "\n",
    "Tools you could use include [BiG-SCAPE](https://bigscape-corason.secondarymetabolites.org/) (slower, sensitive to BGC length biases, easy to interpret output), [CLUST-O-MATIC](https://github.com/Helmholtz-HIPS/clustomatic_source) and [BiG-SLice)(https://github.com/medema-group/bigslice)\n",
    "\n",
    "### BiG-SLiCE: Biosynthetic Gene clusters - Super Linear Clustering Engine\n",
    "See the paper [here](https://academic.oup.com/gigascience/article/10/1/giaa154/6092777)\n",
    "\n",
    "BiG-SLiCE allows clustering thousands of BGCs very quickly and build databases based on the results. It also allows querying your own set of BGCs against a dataset contain 1.2 million BGCs.\n",
    "\n",
    "Here we will use the BiG-FAM database as a refernce to compare the potential novelty of BGCs in our dataset.\n",
    "\n",
    "The reference database containing 1.2 million BGCs is also available online and named (BiG-FAM.)[https://bigfam.bioinformatics.nl/home) Let's open and explore that.\n",
    "\n",
    "Here is a description from the website: \n",
    "\n",
    "    The Biosynthetic Gene Cluster Family (GCF) database is an online repository for \"homologous\" groups of biosynthetic gene clusters (BGCs) putatively encoding the production of similar specialized metabolites.     By taking large-scale, global collections of BGCs identified from currently available genomes and MAGs as a data source, BiG-FAM provides an explorable \"atlas\" of microbial secondary metabolic diversity to b     browse and search biosynthetic diversity across taxa. BiG-FAM facilitates querying putative BGCs to rapidly find their position on the diversity map and gain a better understanding of their novelty or            functions, based on relationships with other known and predicted BGCs from publicly available data.\n",
    "    For the Big-SLiCE analysis, we will create a new directory with all the BGC .gbks from our earlier antiSMASH output. The reference database is already prepared, and named \"full_run_result\"\n",
    "\n",
    "Here it is also important to consider clustering thresholds (how similar gene clusters are to be clustered together). Furthermore, when doing the BiG-SLiCE query analysis the (euclidean) distance between our BGCs and those in the database is important. For the moment, we use the thresholds defined in the paper as distance.\n",
    "\n",
    "### Make sure to speak to us later if you are interested in setting up BiG-SLiCE to query BGCs on our own server or laptop.\n",
    "\n",
    "### Here are the steps for this analysis:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6f4bc6a-6de4-4e65-8b21-1c93d3c769ab",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda activate bigslice\n",
    "cd ../10_BIGSLICE\n",
    "bigslice --query example_input/ --run_id 6 --query_name workshop_12092023 full_run_result/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9685ad3-5426-4fa2-b96f-05436392b596",
   "metadata": {},
   "source": [
    "### BiG-SLiCE results\n",
    "\n",
    "The results are output to a very large database, containing 1.2 million BGCs with their taxonomy and already clustered into gene cluster families (GCFs). These data are stored in a sqlite database, which can house very large amounts of data, but it is not very intuitive to use for new users.\n",
    "\n",
    "On your own server or computer, you can explore the results using the Flask webserver as follows. Instructions can be found [here](https://github.com/medema-group/bigslice/tree/master/misc/input_folder_template)\n",
    "\n",
    "The BiG-SLiCE output provides distance values - distance between your BGCs and the best-matching gene-cluster family in the database. A distance of < 900 is considered \"related\" to the closest reference, a distance > 900 and < 1800 means it is distantly related to the GCF in the reference, and a distance of > 1800 means extremely divergent from any known references. \n",
    "\n",
    "I will provide a python script to extract the necessary information from the BiG-SLiCE analysis, which we will use to make a figure in R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c892f9ab-8ff2-4f42-9a14-0d7bcbe7abf7",
   "metadata": {},
   "outputs": [],
   "source": [
    "wget https://raw.githubusercontent.com/Zander22/metallophore/main/plot_BGC_distances.R\n",
    "wget wget https://raw.githubusercontent.com/Zander22/metallophore/main/extract_distances.py\n",
    "#\n",
    "cd full_run_result\n",
    "#\n",
    "python get_report.py\n",
    "python extract_distances.py\n",
    "#\n",
    "cp exported_full_tables_Sp_Report/merged_table.txt ../\n",
    "#\n",
    "cd ..\n",
    "#\n",
    "conda deactivate\n",
    "conda activate r\n",
    "#\n",
    "Rscript ./plot_BGC_distances.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a87c224d-3923-44ca-a629-d49097d5f7b0",
   "metadata": {},
   "source": [
    "How divergent were the BGCs in this dataset? Do you think there are any that are completely novel?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c65a0eb8-188d-48c8-a6ed-47c510abb2e9",
   "metadata": {},
   "source": [
    "## If time allows, we will now use a different method to cluster BGCs as a means to dereplicate, identify and visualize the presence of difference BGCs across genomes.\n",
    "\n",
    "We will use the [CLUST-O-MATIC](https://github.com/Helmholtz-HIPS/clustomatic_source) tool from the Ziemert lab. \n",
    "\n",
    "The db2fasta.py script will convert our cluster (region) .gbk files to fasta files and then we run clust-o-matic.\n",
    "db2fasta takes directories labelled according to genomes, with the cluster .gbk files inside, and will create a single fasta file.\n",
    "\n",
    "Then running clust-o-matic, the input will be the fasta file created, and the clustering threshold."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a31018de-4557-4fa0-b5d3-055d4c3c82c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "conda deactivate\n",
    "conda activate clustomatic\n",
    "cd ../11_CLUSTOMATIC\n",
    "# first we cluster the BGCs from the genome bins and see whether any BGCs are shared between genome bins\n",
    "python gb2fasta.py input_clusters/ BGCs_genomes.faa\n",
    "python clustomatic.py BGCs_genomes.faa 0.25 > BGCs_genomes_clusters.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17955685-0ec8-4197-9731-3e6faede03e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "head -n 50 BGCs_genomes_clusters.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4df4a53c-473a-4f6e-9b1d-1699e479cf70",
   "metadata": {},
   "source": [
    "You can see in the output the genome name in the first column, and the gene cluster family number in the second column. We will parse these in R and plot a heatmap"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "925ec192-1d5b-4835-8e83-a14cce5276c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# now we cluster our genome bins together with the MiBiG database to see whether this approach allows us to identify some of these clusters\n",
    "python gb2fasta.py input_clusters/ BGCs_mibig.faa\n",
    "python clustomatic.py BGCs_mibig.faa 0.25 > mibig_clusters.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d2e46f0-41bf-4907-8c45-3343410b6895",
   "metadata": {},
   "outputs": [],
   "source": [
    "head -n 50 mibig_clusters.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5f638f3c-9231-40d8-b13b-24b41676c19b",
   "metadata": {},
   "source": [
    "Again as above, but this time MiBiG BGCs are added to the output. We will filter these to any possible hits to your genome bins and make another heatmap in R."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f7c3b774-2c50-46ed-ab6b-13fdb001593a",
   "metadata": {},
   "outputs": [],
   "source": [
    "wget https://raw.githubusercontent.com/Zander22/metallophore/main/heatmap_BGC_genomes.R\n",
    "wget https://raw.githubusercontent.com/Zander22/metallophore/main/heatmap_BGC_mibig.R\n",
    "conda deactivate\n",
    "conda activate r\n",
    "Rscript ./heatmap_BGC_genomes.R\n",
    "Rscript ./heatmap_BGC_mibig.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e74a3e29-d8f6-4bd4-b411-06cfdc927ac4",
   "metadata": {},
   "source": [
    "## A few tools to further explore the function, evolution or relevance of your BGCs\n",
    "\n",
    "[ARTS:](http://arts.ziemertlab.com/) Antibiotic Resistant Target Seeker, a tool to help identify resistance genes within BGCs.\r",
    "[\n",
    "NaPDoS:](https://npdomainseeker.sdsc.edu/napdos2/napdos_home_v2.html)2 classification of ketosynthase (KS) and condensation (C) domains[\r\n",
    "EvoMini:](https://github.com/nselem/EvoMining/wiki)ng genome-mining and visualization approach for biosynthetic enzyme discovery that incorporates evolutionary principle[s\r\n",
    "AutoM:](https://automlst.ziemertlab.com/)LST automatic generation of a species phylogeny with reference organi\n",
    "\n",
    "ARTS is a really cool tool, and could potentially aid you in identifying whether some of your novel BGCs code for antimicrobial compounds, based on the assumption that BGCs will have antibiotic resistance genes location on or near them. If there is time. Open the link...sms"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce38e77b-d794-4ec1-91e5-c5b978f068af",
   "metadata": {},
   "source": [
    "## Summary\n",
    "\n",
    "You have:\n",
    " - mined your genome bins for BGCs and explored its output\n",
    " - clustered your BGCs to a database with 1.2 million BGCs and determined whether any of them are novel clusters\n",
    " - clustered your BGCs to try and identify them, and dereplicate them in your dataset."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
